### ISSUE TITLE
False positive unreachable when pattern matching `int` with `http.HTTPStatus`

### ISSUE THREAD

**Issue by injust** [microsoft/pyright#10235]
**Describe the bug**

When using structural pattern matching to compare an `int` vs an `http.HTTPStatus` instance (and vice versa), pyright thinks the code is unreachable.

In the example code, pyright thinks the wildcard case is taken in both scenarios (hence the lack of output).

**Code or Screenshots**

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBAFgLnADlAlhRYRygCQCp4AKAynAIZwCuAzgFCiRRwCeiKAdgOarqbYgBTAG4CyAGwD6LRANq0AxmAAmAqAF4oAJgAM2uRAryYURSoBctKFZNlqq-EVIUaACgCUF616iJbdL-K29gQk5FTUAHQA8gDSnt5WgiLiUqwCLqYCbpbWgXZQEvEJvtR0ctRhNOq4IU7h0TH6hsYVztRFNvk62h1eSaKS0umt4dkBQQW91iV0QA)

```python
from http import HTTPStatus
from typing import reveal_type

code = 200

match code:
    case HTTPStatus():
        pass
    case HTTPStatus.OK:
        reveal_type(code)
    case _:
        pass


status = HTTPStatus.OK

match status:
    case 200:
        reveal_type(status)
    case _:
        pass

```

**Comment by erictraut** [comment]
Literals require a bunch of special-cased type narrowing logic. The `Enum` class also requires a bunch of special casing because its metaclass performs so much non-standard "magic" under the covers. `IntEnum` and `StrEnum` involve both of these two concepts, so they require special cases on top of special cases. I'll see if I can improve this without adding too much added complexity.
